### ISSUE TITLE
Exact same training config gives slightly different results in different runs

### ISSUE THREAD

**Issue by rihardsk** [marian-nmt/marian-dev#609]
This isn't really an issue in any practical scenarios but it's still a bit weird and i'm curious to know what causes it.

As the title suggests, i tried running the same training configuration two times (to be precise - on two gpus in parallel), and got slightly different results. Training and validation loss scores differ by ~0.1. Validation BLEU scores differ by ~0.4 BLEU.

Here are the full logs - [train-1.log](https://github.com/marian-nmt/marian-dev/files/4260589/train-1.log), [train-2.log](https://github.com/marian-nmt/marian-dev/files/4260591/train-2.log). I'm using a forked version of Marian with some modifications and it's a few months behind master but it shouldn't be the cause of the nondeterminism. If anyone's interested, i can re-run on the latest master.

**Comment by snukky** [comment]
I've noticed a similar nondeterminism while developing regression tests and a solution was to set eps to 0 in the Adam optimizer. It has been (accidentally?) fixed in the current marian-dev quite recently. Of course, I don't know if this is related to your case, and it seems you use your own optimizer.

Marian has changed quite a lot since 1.7.6, which has been released 15 months ago, including minor bug fixes in the backend.

**Comment by rihardsk** [comment]
I repeated the experiment using adam with the latest Marian. There's still nondeterminism. This time, though, i used two different gpus (didn't have anything else available this time, hoped it doesn't matter, does it?).

Here are the logs [train1.log](https://github.com/marian-nmt/marian-dev/files/4287458/train1.log), [train2.log](https://github.com/marian-nmt/marian-dev/files/4287461/train2.log)

**Comment by emjotde** [comment]
Different GPU generations? That will certainly matter and likely result in different computation.

There is never a determism guarantee for GPU for training even on the same GPU as long as we have to use atomicAdd for a few kernels. This causes floating point number addition in slightly different order due to multi-threading serialization and since floating point addition is not associative you get non-determinism. That does not mean that there isn't a bug that exacerbates that.

**Comment by rihardsk** [comment]
Thanks for the explanation.

Yup, the latest logs were from GPUs of different generations. I reran the experiment on identical GPUs - now the valid/train loss/BLEU scores match perfectly. Closing.
